### PR TITLE
Precondition ChannelArgs with EventEngines

### PR DESCRIPTION
**Pull request by drfloob**
If an EventEngine is not explicitly provided to ChannelArgs, the default EventEngine will be set when ChannelArgs are preconditioned.

This relands the second half of #30556, along with some added engine source migrations. shared_ptr support was added in https://github.com/grpc/grpc/pull/31056

### CODE REVIEW

File: src/core/lib/event_engine/default_event_engine.cc
Comment (markdroth) [inline src/core/lib/event_engine/default_event_engine.cc:92]: Suggest calling this `RegisterEventEngineChannelArgPreconditioning()`, to make it clear what it's actually registering.

**Review by markdroth — COMMENTED** [review]
This looks good overall!

Please let me know if you have any questions.  Thanks!

File: include/grpc/impl/codegen/grpc_types.h
Comment (drfloob) [reply include/grpc/impl/codegen/grpc_types.h]: Done

File: src/core/lib/event_engine/default_event_engine.cc
Comment (drfloob) [reply src/core/lib/event_engine/default_event_engine.cc]: Done

File: src/core/ext/filters/client_channel/lb_policy/weighted_target/weighted_target.cc
Comment (drfloob) [reply src/core/ext/filters/client_channel/lb_policy/weighted_target/weighted_target.cc]: Sounds good, let's chat about channel-wide mechanisms soon. For now, I've made the change to have the WeightedTargetLb own an engine ref (taken from ChannelArgs on lb policy construction).

File: src/core/ext/filters/client_channel/lb_policy/weighted_target/weighted_target.cc
Comment (markdroth) [inline src/core/ext/filters/client_channel/lb_policy/weighted_target/weighted_target.cc:178]: This method is not necessary.  `DelayedRemovalTimer` is nested inside of `WeightedChild` and can therefore access its private data members directly.

File: src/core/ext/filters/client_channel/lb_policy/xds/xds_cluster_manager.cc
Comment (markdroth) [inline src/core/ext/filters/client_channel/lb_policy/xds/xds_cluster_manager.cc:465]: This LB policy should use the same pattern as the grpclb and weighted_target LB policies: The LB policy should be given a `std::shared_ptr<EventEngine>` data member that will be set in its ctor using `GetObjectRef<>()`, and the nested classes can access that data member directly instead of fishing it out of channel args themselves.

File: src/core/ext/filters/client_channel/lb_policy/weighted_target/weighted_target.cc
Comment (drfloob) [reply src/core/ext/filters/client_channel/lb_policy/weighted_target/weighted_target.cc]: Fixed, thanks.

File: src/core/ext/filters/client_channel/lb_policy/xds/xds_cluster_manager.cc
Comment (drfloob) [reply src/core/ext/filters/client_channel/lb_policy/xds/xds_cluster_manager.cc]: Good call, done. I've also renamed the Subchannel member to keep things consistent.

**Comment by ctiller** [comment]
I'm not sure this will be safely usable yet: we need to have some story with init.cc and the reset of event engine we do there prior to shutdown. If the event engine is not quiesced before iomgr shuts down, we have a lot of code that assumes it will be available that will not be callable from `EventEngine::Run` and friends.

**Comment by drfloob** [comment]
Thanks for pointing that out, @ctiller. ResetDefaultEventEngine isn't necessary anymore now that gRPC keeps a weak_ptr instead of a raw pointer, so I've removed the function https://github.com/grpc/grpc/pull/31213. To clarify the story, though, this means that new EventEngines will be created as necessary, even after gRPC is shut down (for however long that concept exists). And gRPC internals will not keep those engines alive if the callers of `GetDefaultEventEngine` let go of their references.

**Comment by drfloob** [comment]
I believe that means this PR is safe to land. Please chime in if you disagree.

**Comment by ctiller** [comment]
This nor #31213 are safe to land right now. We definitely have code that assumes that it can access iomgr related code that can be triggered via an `EventEngine::Run` callback.

**Comment by ctiller** [comment]
(see #31208 for a writeup on some of the stuff happening here)

**Comment by markdroth** [comment]
Craig's #31265 should allow us to move forward with most of this, but I think we want to do it a little differently.

I've sent https://github.com/drfloob/grpc/pull/333 against this PR with what I think is the approach we want here.  (Note that most of the changes there are from merging master; the actual changes that I made are in [the last 5 commits](https://github.com/drfloob/grpc/pull/333/files/f71af7dceb34b35d8dbd0eee39906816b9fa1cef..5e0ac029aa2a5d694f0caaaa526eacc308277f44)).

The main idea here is that we want most of the modules within C-core to get the EE instance from a central location in the channel rather than storing their own refs to the EE instance.  For now, that central location is the `grpc_channel_stack` struct (which will eventually be merged into `grpc_core::Channel`).  Both `ChannelIdleFilter` and `ClientChannel` are now getting the EE instance from that central point.  And I've added a new method to the LB helper interface for the LB policies to use, so that they don't need to dig it out of channel args or store their own refs either.

Note that using `grpc_channel_stack` as the central point to store the EE instance makes sense, since we probably do want to hold our own ref to the EE instance in both the parent channel and in each subchannel stack.  However, it also means that we're storing another ref in the dynamic filter stack, which is not strictly necessary.  But we can work on ways to optimize out some redundancy later as needed (and I suspect some of that will naturally fall out of the promise conversion).

Also, we don't yet want to allow passing in the EE instance from the application via channel args, so I've removed the channel arg macro from the include tree and changed the string to start with `grpc.internal.`, so that it cannot be passed in from outside.  We are still using the infrastructure from this PR to pass the EE instance around via channel args *within* C-core; we're only disallowing passing it in from outside.  This will change as we get further along with eliminating the iomgr code.

Note that we are still using channel args to pass the EE instance down into the `Subchannel` object itself, which is still storing its own ref to the EE instance.  That is actually a bit sub-optimal, since this effectively duplicates the ref that will already be held in the subchannel stack.  However, the subchannel stack exists only when the subchannel is actually connected, and it needs to schedule the backoff timer when it is not connected, so I don't see any easy way to avoid this duplication for now.

Finally, Craig tells me that with #31265, there are deadlock issues if code within iomgr itself tries to instantiate an EE instance, so I've reverted the changes resolve_address_posix.{h,cc}.  I think it's fine for iomgr to use the default global EE instance, since iomgr is global anyway, not per-channel.

@ctiller and @drfloob, please review my changes and let me know what you think.  Thanks!

**Comment by drfloob** [comment]
Thanks Mark! I like this approach, and I've merged it in.

> Also, we don't yet want to allow passing in the EE instance from the application via channel args

For transparency, the main reason is to simplify our debugging as we continue to land EventEngine usage in core. Having multiple EventEngines in play adds complexity, which we can defer to make the migration easier. There's nothing technically preventing us from exposing the API at this point. If there are sufficiently motivating use cases to open the API, we will do so.

File: test/core/channel/channel_stack_test.cc
Comment (markdroth) [inline test/core/channel/channel_stack_test.cc:39]: If we need to use this in tests, I'd be fine with adding such a macro, as long as it's in a file in src instead of in include.

File: test/core/filters/filter_fuzzer.cc
Comment (markdroth) [inline test/core/filters/filter_fuzzer.cc:251]: Should we just precondition the channel args here?  If these args are going to be passed to a filter, the filter will assume that they've already been preconditioned.

File: BUILD
Comment (markdroth) [inline BUILD:1762]: This dependency should not be needed.  Preconditioning should be done inside of C-core, not at the C++ layer.

File: BUILD
Comment (markdroth) [inline BUILD:1823]: Same here.

File: src/core/lib/channel/channel_args_preconditioning.h
Comment (markdroth) [inline src/core/lib/channel/channel_args_preconditioning.h:57]: I don't think we should need this override.  The only places where we should be preconditioning channel args are at the C-core API boundary, where the caller passes in C-style args and we return `grpc_core::ChannelArgs`.

File: BUILD
Comment (drfloob) [reply BUILD]: Done

File: BUILD
Comment (drfloob) [reply BUILD]: Done

File: test/core/filters/filter_fuzzer.cc
Comment (drfloob) [reply test/core/filters/filter_fuzzer.cc]: Done

File: test/core/channel/channel_stack_test.cc
Comment (drfloob) [reply test/core/channel/channel_stack_test.cc:40]: Added in channel_args.h as `GRPC_INTERNAL_ARG_EVENT_ENGINE`

File: src/core/ext/filters/client_channel/lb_policy/xds/xds_wrr_locality.cc
Comment (drfloob) [reply src/core/ext/filters/client_channel/lb_policy/xds/xds_wrr_locality.cc]: Done

File: src/core/lib/channel/channel_args_preconditioning.h
Comment (drfloob) [reply src/core/lib/channel/channel_args_preconditioning.h:57]: See my comment above regarding wherever we create ChannelArgs internally. There are exceptions where it's necessary, and I didn't think it would be right to backtrack to using the C-type just for this. 

This gave me an idea, however: We could instead ensure that all ChannelArgs are preconditioned in the constructor.

File: src/core/lib/channel/channel_stack.cc
Comment (markdroth) [reply src/core/lib/channel/channel_stack.cc:122]: In principal, I think this preconditioning should be done where the channel args are created, not where they are used.  How many places do we create new channel args within C-core?

We should also consider why we are creating channel args within C-core in the first place.  In principal, there should be very few places where this happens -- it should generally be necessary only when constructing a control-plane channel within C-core, such as the xDS channel, the grpclb channel, or the ALTS channel, and those places should get preconditioning automatically via `grpc_channel_create()`.  If there are other places where we're doing this, I think we should look at each one and decide whether it's actually necessary.

For example, do we actually have any cases where we use `ChannelStackBuilder` without calling `SetChannelArgs()`?  If not, maybe we should just change the `ChannelStackBuilder` API to require passing in channel args to its ctor, so that this case goes away.

I don't think we want the preconditioning to happen whenever we instantiate a `ChannelArgs` object, because those objects get copied all the time within C-core.

Comment (drfloob) [reply src/core/lib/channel/channel_stack.cc]: > How many places do we create new channel args within C-core?

To begin with, I count around 30 default-constructed ChannelArgs class members within src/core. I haven't gone through to see how many get overwritten, or could otherwise be accessed before being overwritten - LB policies and client channel resolvers make up about a third of the set. I'll take a deeper look here.

I did the experiment you suggested with ChannelStackBuilder, indeed we can add a ChannelArgs argument to the constructor, and all PR tests pass. So we may be safe with that change alone.

> I don't think we want the preconditioning to happen whenever we instantiate a ChannelArgs object, because those objects get copied all the time within C-core.

... and we'd be adding a dependency cycle between core config and channel args, as well as making making channel args fragments more expensive (as used in resolvers). It was a thought, but wouldn't pan out right now.

File: src/core/lib/channel/channel_args_preconditioning.h
Comment (drfloob) [reply src/core/lib/channel/channel_args_preconditioning.h:57]: I removed this overload. I have a hunch we may still need it, but I was able to remove the need for it in this PR. Barring discovery that any internally-created ChannelArgs need preconditioning, I think it can stay gone.

File: src/core/lib/channel/channel_stack_builder.h
Comment (markdroth) [inline src/core/lib/channel/channel_stack_builder.h:68]: Do we still call this anywhere?  If not, we can remove it.

File: src/core/lib/channel/channel_args_preconditioning.h
Comment (markdroth) [reply src/core/lib/channel/channel_args_preconditioning.h:57]: Looks like you removed the definition of the overload in the .cc file but forgot to remove the declaration here in the .h file.

**Review by markdroth — APPROVED** [review]
This looks good to me, modulo a couple of minor comments below.

Please make sure @ctiller reviews as well before merging.

Thanks!

File: src/core/lib/channel/channel_stack_builder.h
Comment (drfloob) [reply src/core/lib/channel/channel_stack_builder.h]: It was used in one place, but I was able to refactor that usage away without much extra spaghettification.

**Comment by drfloob** [comment]
> Looks like you removed the definition of the overload in the .cc file but forgot to remove the declaration here in the .h file.
Good catch, removed.

File: src/core/lib/channel/channel_args_preconditioning.h
Comment (drfloob) [reply src/core/lib/channel/channel_args_preconditioning.h]: Removed, thanks.

File: src/core/lib/surface/channel.h
Comment (ctiller) [inline src/core/lib/surface/channel.h:159]: (no change required)

Probably ok for now, we might want to cache the pointer on this object at some point though (thinking about number of pointer chases on common per-call paths)

CC @markdroth

**Review by ctiller — APPROVED** [review]
looks good... left one comment that needs no action (but wanted to note)

File: src/core/lib/surface/channel.h
Comment (markdroth) [reply src/core/lib/surface/channel.h:159]: I was thinking about that, but wouln't any per-call paths get this pointer from the promise context instead?  I would expect that only control-plane code would actually wind up using this method.

**Pull request merged**
